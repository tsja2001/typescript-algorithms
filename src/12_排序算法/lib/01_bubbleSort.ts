import { swap } from './utils'

export function bubbleSort(arr: number[]): number[] {
  const n = arr.length

  let swaped = false

  for (let i = 0; i < n; i++) {
    for (let j = 0; j < n - 1 - i; j++) {
      if (arr[j] > arr[j + 1]) {
        swap(arr, j, j + 1)
        swaped = true
      }
    }
		
    if (!swaped) {
      break
    }
  }

  return arr
}

const arr = [9, 5, 8, 3, 1, 0]

const sortArr = bubbleSort(arr)

console.log(sortArr)
