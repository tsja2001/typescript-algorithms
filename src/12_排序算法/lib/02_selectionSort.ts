import { measureSort } from 'hy-algokit'
import { swap } from './utils'

export function selectionSort(arr: number[]): number[] {
  const n = arr.length
  for (let i = 0; i < n - 1; i++) {
    let minIndex = i
    for (let j = i + 1; j < n; j++) {
      if (arr[j] < arr[minIndex]) {
        minIndex = j
      }
    }

		// 优化: 如果当前元素不需要移动, 则不进行交换
    if (i !== minIndex) {
      swap(arr, i, minIndex)
    }
  }

  return arr
}
const arr = [9, 5, 8, 3, 1, 0]

const sortArr = selectionSort(arr)

console.log(sortArr)
measureSort(selectionSort)
