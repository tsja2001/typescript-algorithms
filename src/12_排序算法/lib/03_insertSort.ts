export const insertionSort = (arr: number[]) => {
  const n = arr.length
  for (let i = 1; i < n; i++) {
		const newNum = arr[i]
		let j = i - 1
    while (arr[j] > newNum  && j >= 0) {
			arr[j + 1] = arr[j]
      j--
    }
		arr[j + 1] = newNum
  }

  return arr
}
const arr = [9, 5, 8, 3, 1, 0]

const sortArr = insertionSort(arr)

console.log(sortArr)
