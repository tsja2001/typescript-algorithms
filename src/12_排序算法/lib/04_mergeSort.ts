export function mergeSort(arr: number[]): number[] {
  if (arr.length <= 1) return arr

  const middle = Math.floor(arr.length / 2)
  const leftArr = arr.slice(0, middle)
  const rightArr = arr.slice(middle)

  const lastLeftArr = mergeSort(leftArr)
  const lastRightArr = mergeSort(rightArr)

  let i = 0
  let j = 0

  const resArr = []

  while (i < lastLeftArr.length && j < lastRightArr.length) {
    if (lastLeftArr[i] < lastRightArr[j]) {
      resArr.push(lastLeftArr[i])
      i++
    } else {
      resArr.push(lastRightArr[j])
      j++
    }
  }

  if (i < lastLeftArr.length) {
    resArr.push(...lastLeftArr.slice(i))
  }
  if (j < lastRightArr.length) {
    resArr.push(...lastRightArr.slice(j))
  }

	return resArr
}

const arr = [9, 5, 8, 3, 1, 0]
debugger

const sortArr = mergeSort(arr)

console.log(sortArr)

