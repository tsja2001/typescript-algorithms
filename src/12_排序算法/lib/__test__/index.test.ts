import fs from 'fs'
import { bubbleSort } from '../01_bubbleSort'
import { selectionSort } from '../02_selectionSort'
import { insertionSort } from '../03_insertSort'
import { mergeSort } from '../04_mergeSort'
import path from 'path'

function shuffle<T>(array: T[]): T[] {
  const result = [...array]
  for (let i = result.length - 1; i > 0; i--) {
    const j = Math.floor(Math.random() * (i + 1))
    ;[result[i], result[j]] = [result[j], result[i]]
  }
  return result
}

describe('Sorting algorithms', () => {
  const testCases = [
    {
      description: 'empty array',
      input: [],
      expected: []
    },
    {
      description: 'array with one element',
      input: [1],
      expected: [1]
    },
    {
      description: 'array in descending order',
      input: [3, 2, 1],
      expected: [1, 2, 3]
    },
    {
      description: 'array with repeated elements',
      input: [5, 2, 7, 2, 8, 6],
      expected: [2, 2, 5, 6, 7, 8]
    },
    {
      description: 'array in ascending order',
      input: [1, 2, 3, 4, 5],
      expected: [1, 2, 3, 4, 5]
    },
    {
      description: 'array with negative numbers',
      input: [5, -2, 7, 3, -8, 6],
      expected: [-8, -2, 3, 5, 6, 7]
    },
    {
      description: 'array with large number of elements',
      input: shuffle(Array.from({ length: 9999 }, (_, i) => i)),
      expected: Array.from({ length: 9999 }, (_, i) => i)
    },
    {
      description:
        'array with large number of elements in descending order',
      input: Array.from({ length: 9999 }, (_, i) => 9998 - i),
      expected: Array.from({ length: 9999 }, (_, i) => i)
    }
  ]

  const results: Record<string, Record<string, any>> = {}

  test.each`
    sortFn           | functionName
    ${bubbleSort}    | ${'bubble sort'}
    ${selectionSort} | ${'selection sort'}
    ${insertionSort} | ${'insertion sort'}
    ${mergeSort}     | ${'merge sort'}
  `('$functionName', ({ sortFn, functionName }) => {
    results[functionName] = {}
    testCases.forEach(({ description, input, expected }) => {
      const start = performance.now()
      sortFn(input)
      const end = performance.now()
      const duration = end - start
      results[functionName][description] = duration * 1000
      expect(sortFn(input)).toEqual(expected)
    })
    const durations = Object.values(results[functionName])
    const avgDuration =
      durations.reduce((sum, duration) => sum + duration, 0) /
      durations.length
    results[functionName]['average'] = avgDuration
  })

  afterAll(() => {
    const filePath = path.resolve(__dirname, './results.json')
    const output = { algorithms: results }
    fs.writeFile(filePath, JSON.stringify(output, null, 2), (err) => {
      if (err) throw err
      console.log('Results saved to file!')
    })
  })
})
