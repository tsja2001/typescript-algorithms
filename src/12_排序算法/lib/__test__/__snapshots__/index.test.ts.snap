// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Sorting algorithms bubble sort 1`] = `[]`;

exports[`Sorting algorithms bubble sort 2`] = `
[
  1,
]
`;

exports[`Sorting algorithms bubble sort 3`] = `
[
  1,
  2,
  3,
]
`;

exports[`Sorting algorithms bubble sort 4`] = `
[
  2,
  3,
  5,
  6,
  7,
  8,
]
`;

exports[`Sorting algorithms bubble sort 5`] = `
[
  2,
  2,
  5,
  6,
  7,
  8,
]
`;

exports[`Sorting algorithms bubble sort 6`] = `
[
  1,
  2,
  3,
  4,
  5,
]
`;

exports[`Sorting algorithms bubble sort 7`] = `
[
  1,
  2,
  3,
  4,
  5,
]
`;

exports[`Sorting algorithms bubble sort 8`] = `
[
  -8,
  -2,
  3,
  5,
  6,
  7,
]
`;

exports[`Sorting algorithms insertion sort 1`] = `[]`;

exports[`Sorting algorithms insertion sort 2`] = `
[
  1,
]
`;

exports[`Sorting algorithms insertion sort 3`] = `
[
  1,
  2,
  3,
]
`;

exports[`Sorting algorithms insertion sort 4`] = `
[
  2,
  3,
  5,
  6,
  7,
  8,
]
`;

exports[`Sorting algorithms insertion sort 5`] = `
[
  2,
  2,
  5,
  6,
  7,
  8,
]
`;

exports[`Sorting algorithms insertion sort 6`] = `
[
  1,
  2,
  3,
  4,
  5,
]
`;

exports[`Sorting algorithms insertion sort 7`] = `
[
  1,
  2,
  3,
  4,
  5,
]
`;

exports[`Sorting algorithms insertion sort 8`] = `
[
  -8,
  -2,
  3,
  5,
  6,
  7,
]
`;

exports[`Sorting algorithms merge sort 1`] = `[]`;

exports[`Sorting algorithms merge sort 2`] = `
[
  1,
]
`;

exports[`Sorting algorithms merge sort 3`] = `
[
  1,
  2,
  3,
]
`;

exports[`Sorting algorithms merge sort 4`] = `
[
  2,
  3,
  5,
  6,
  7,
  8,
]
`;

exports[`Sorting algorithms merge sort 5`] = `
[
  2,
  2,
  5,
  6,
  7,
  8,
]
`;

exports[`Sorting algorithms merge sort 6`] = `
[
  1,
  2,
  3,
  4,
  5,
]
`;

exports[`Sorting algorithms merge sort 7`] = `
[
  1,
  2,
  3,
  4,
  5,
]
`;

exports[`Sorting algorithms merge sort 8`] = `
[
  -8,
  -2,
  3,
  5,
  6,
  7,
]
`;

exports[`Sorting algorithms selection sort 1`] = `[]`;

exports[`Sorting algorithms selection sort 2`] = `
[
  1,
]
`;

exports[`Sorting algorithms selection sort 3`] = `
[
  1,
  2,
  3,
]
`;

exports[`Sorting algorithms selection sort 4`] = `
[
  2,
  3,
  5,
  6,
  7,
  8,
]
`;

exports[`Sorting algorithms selection sort 5`] = `
[
  2,
  2,
  5,
  6,
  7,
  8,
]
`;

exports[`Sorting algorithms selection sort 6`] = `
[
  1,
  2,
  3,
  4,
  5,
]
`;

exports[`Sorting algorithms selection sort 7`] = `
[
  1,
  2,
  3,
  4,
  5,
]
`;

exports[`Sorting algorithms selection sort 8`] = `
[
  -8,
  -2,
  3,
  5,
  6,
  7,
]
`;
